/*
** Copyright 2022 Centreon
**
** Licensed under the Apache License, Version 2.0 (the "License");
** you may not use this file except in compliance with the License.
** You may obtain a copy of the License at
**
**     http://www.apache.org/licenses/LICENSE-2.0
**
** Unless required by applicable law or agreed to in writing, software
** distributed under the License is distributed on an "AS IS" BASIS,
** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
** See the License for the specific language governing permissions and
** limitations under the License.
**
** For more information : contact@centreon.com
*/

syntax = "proto3";

package com.centreon.engine.configuration;

message CustomVariable {
  string name = 1;
  string value = 2;
  bool is_sent = 3;
  bool modified = 4;
}

message Object {
  string name = 1;
  bool register = 2;
  repeated string use = 3;
}

message Point2d {
  int32 x = 1;
  int32 y = 2;
}

message Point3d {
  double x = 1;
  double y = 2;
  double z = 3;
}

message KeyType {
  uint64 id = 1;
  uint32 type = 2;
}

message DaysArray {
  string sunday = 1;
  string monday = 2;
  string tuesday = 3;
  string wednesday = 4;
  string thursday = 5;
  string friday = 6;
  string saturday = 7;
}

message TimeRange {
  uint64 range_start = 1;
  uint64 range_end = 2;
}

message Daterange {
  enum TypeRange {
    calendar_date = 0;
    month_date = 1;
    month_day = 2;
    month_week_day = 3;
    week_day = 4;
    none = 5; // Instead of -1 in original config
  }
  TypeRange type = 1;
  int32 syear = 2;  // Start year.
  int32 smon = 3;   // Start month.
  // Start day of month (may 3rd, last day in feb).
  int32 smday = 4;
  int32 swday = 5;  // Start day of week (thursday).
  // Start weekday offset (3rd thursday, last monday in jan).
  int32 swday_offset = 6;
  int32 eyear = 7;
  int32 emon = 8;
  int32 emday = 9;
  int32 ewday = 10;
  int32 ewday_offset = 11;
  int32 skip_interval = 12;
  repeated TimeRange timerange = 13;
}

message ExceptionArray {
  repeated Daterange calendar_date = 1;
  repeated Daterange month_date = 2;
  repeated Daterange month_day = 3;
  repeated Daterange month_week_day = 4;
  repeated Daterange week_day = 5;
}

message PairString {
  string first = 1;
  string second = 2;
}


enum DependencyKind {
  unknown = 0;
  notification_dependency = 1;
  execution_dependency = 2;
}

message Anomalydetection {
  Object obj = 1;
  optional int32 acknowledgement_timeout = 2;  // Optional - Default value: 0
  string action_url = 3;
  optional bool status_change = 4;  // Optional - Default value: false
  optional bool checks_active = 5;  // Optional - Default value: true
  optional bool checks_passive = 6;  // Optional - Default value: true
  string metric_name = 7;
  string thresholds_file = 8;
  optional bool check_freshness = 9;  // Optional - Default value: 0
  optional uint32 check_interval = 10;  // Optional - Default value: 5
  repeated string contactgroups = 11;
  repeated string contacts = 12;
  repeated CustomVariable customvariables = 13;
  string display_name = 14;
  string event_handler = 15;
  optional bool event_handler_enabled = 16;  // Optional - Default value: true
  optional uint32 first_notification_delay = 17;  // Optional - Default value: 0
  optional bool flap_detection_enabled = 18;  // Optional - Default value: true
  optional uint32 flap_detection_options = 19;  // Optional
  optional uint32 freshness_threshold = 20;  // Optional - Default value: 0
  optional uint32 high_flap_threshold = 21;  // Optional - Default value: 0
  string host_name = 22;
  string icon_image = 23;
  string icon_image_alt = 24;
  optional uint32 initial_state = 25;  // Optional - Default value: engine::service::state_ok
  optional bool is_volatile = 26;  // Optional - Default value: false
  optional uint32 low_flap_threshold = 27;  // Optional - Default value: 0
  optional uint32 max_check_attempts = 28;  // Optional - Default value: 3
  string notes = 29;
  string notes_url = 30;
  optional bool notifications_enabled = 31;  // Optional - Default value: true
  optional uint32 notification_interval = 32;  // Optional - Default value: 30
  optional uint32 notification_options = 33;  // Optional
  optional string notification_period = 34;  // Optional
  optional bool obsess_over_service = 35;  // Optional - Default value: true
  optional bool process_perf_data = 36;  // Optional - Default value: true
  optional bool retain_nonstatus_information = 37;  // Optional - Default value: true
  optional bool retain_status_information = 38;  // Optional - Default value: true
  optional uint32 retry_interval = 39;  // Optional - Default value: 1
  optional uint32 recovery_notification_delay = 40;  // Optional
  repeated string servicegroups = 41;
  string service_description = 42;
  uint64 host_id = 43;
  uint64 service_id = 44;
  uint64 internal_id = 45;
  uint64 dependent_service_id = 46;
  optional uint32 stalking_options = 47;  // Optional - Default value: anomalydetection::none
  optional string timezone = 48;  // Optional
  optional uint64 severity_id = 49;  // Optional
  optional uint64 icon_id = 50;  // Optional
  optional double sensitivity = 51;  // Optional
}

message Command {
  Object obj = 1;
  string command_line = 2;
  string command_name = 3;
  string connector = 4;
}

message Connector {
  Object obj = 1;
  string connector_line = 2;
  string connector_name = 3;
}

message Contact {
  Object obj = 1;
  repeated string address = 2;
  string alias = 3;
  optional bool can_submit_commands = 4;  // Optional - Default value: true
  repeated string contactgroups = 5;
  string contact_name = 6;
  repeated CustomVariable customvariables = 7;
  string email = 8;
  optional bool host_notifications_enabled = 9;  // Optional - Default value: true
  repeated string host_notification_commands = 10;
  optional uint32 host_notification_options = 11;  // Optional - Default value: host::none
  string host_notification_period = 12;
  optional bool retain_nonstatus_information = 13;  // Optional - Default value: true
  optional bool retain_status_information = 14;  // Optional - Default value: true
  string pager = 15;
  repeated string service_notification_commands = 16;
  optional uint32 service_notification_options = 17;  // Optional - Default value: service::none
  string service_notification_period = 18;
  optional bool service_notifications_enabled = 19;  // Optional - Default value: true
  optional string timezone = 20;  // Optional
}

message Contactgroup {
  Object obj = 1;
  string alias = 2;
  repeated string contactgroup_members = 3;
  string contactgroup_name = 4;
  repeated string members = 5;
}

message Host {
  Object obj = 1;
  optional int32 acknowledgement_timeout = 2;  // Optional
  string action_url = 3;
  string address = 4;
  string alias = 5;
  optional bool checks_active = 6;  // Optional - Default value: true
  optional bool checks_passive = 7;  // Optional - Default value: true
  string check_command = 8;
  optional bool check_freshness = 9;  // Optional - Default value: false
  optional uint32 check_interval = 10;  // Optional - Default value: 5
  string check_period = 11;
  repeated string contactgroups = 12;
  repeated string contacts = 13;
  optional Point2d coords_2d = 14;  // Optional
  optional Point3d coords_3d = 15;  // Optional
  repeated CustomVariable customvariables = 16;
  string display_name = 17;
  string event_handler = 18;
  optional bool event_handler_enabled = 19;  // Optional - Default value: true
  optional uint32 first_notification_delay = 20;  // Optional - Default value: 0
  optional bool flap_detection_enabled = 21;  // Optional - Default value: true
  optional uint32 flap_detection_options = 22;  // Optional
  optional uint32 freshness_threshold = 23;  // Optional - Default value: 0
  optional uint32 high_flap_threshold = 24;  // Optional - Default value: 0
  repeated string hostgroups = 25;
  uint64 host_id = 26;
  string host_name = 27;
  string icon_image = 28;
  string icon_image_alt = 29;
  optional uint32 initial_state = 30;  // Optional - Default value: engine::host::state_up
  optional uint32 low_flap_threshold = 31;  // Optional - Default value: 0
  optional uint32 max_check_attempts = 32;  // Optional - Default value: 3
  string notes = 33;
  string notes_url = 34;
  optional bool notifications_enabled = 35;  // Optional - Default value: true
  optional uint32 notification_interval = 36;  // Optional - Default value: 30
  optional uint32 notification_options = 37;  // Optional
  string notification_period = 38;
  optional bool obsess_over_host = 39;  // Optional - Default value: true
  repeated string parents = 40;
  optional bool process_perf_data = 41;  // Optional - Default value: true
  optional bool retain_nonstatus_information = 42;  // Optional - Default value: true
  optional bool retain_status_information = 43;  // Optional - Default value: true
  optional uint32 retry_interval = 44;  // Optional - Default value: 1
  optional uint32 recovery_notification_delay = 45;  // Optional
  optional uint32 stalking_options = 46;  // Optional - Default value: host::none
  string statusmap_image = 47;
  optional string timezone = 48;  // Optional
  string vrml_image = 49;
  optional uint64 severity_id = 50;  // Optional
  optional uint64 icon_id = 51;  // Optional
}

message Hostdependency {
  Object obj = 1;
  string dependency_period = 2;
  DependencyKind dependency_type = 3;
  repeated string dependent_hostgroups = 4;
  repeated string dependent_hosts = 5;
  optional uint32 execution_failure_options = 6;  // Optional
  repeated string hostgroups = 7;
  repeated string hosts = 8;
  optional bool inherits_parent = 9;  // Optional - Default value: false
  optional uint32 notification_failure_options = 10;  // Optional
}

message Hostescalation {
  Object obj = 1;
  repeated string contactgroups = 2;
  optional uint32 escalation_options = 3;  // Optional - Default value: hostescalation::none
  optional string escalation_period = 4;  // Optional
  optional uint32 first_notification = 5;  // Optional - Default value: -2
  repeated string hostgroups = 6;
  repeated string hosts = 7;
  optional uint32 last_notification = 8;  // Optional - Default value: -2
  optional uint32 notification_interval = 9;  // Optional - Default value: 0
}

message Hostgroup {
  Object obj = 1;
  string action_url = 2;
  string alias = 3;
  uint32 hostgroup_id = 4;
  string hostgroup_name = 5;
  repeated string members = 6;
  string notes = 7;
  string notes_url = 8;
}

message Hostextinfo {
  Object obj = 1;
  string action_url = 2;
  optional Point2d coords_2d = 3;  // Optional - Default value: -1, -1
  optional Point3d coords_3d = 4;  // Optional - Default value: 0.0, 0.0, 0.0
  repeated string hostgroups = 5;
  repeated string hosts = 6;
  string icon_image = 7;
  string icon_image_alt = 8;
  string notes = 9;
  string notes_url = 10;
  string statusmap_image = 11;
  string vrml_image = 12;
}

message Service {
  Object obj = 1;
  optional int32 acknowledgement_timeout = 2;  // Optional - Default value: 0
  string action_url = 3;
  optional bool checks_active = 4;  // Optional - Default value: true
  optional bool checks_passive = 5;  // Optional - Default value: true
  string check_command = 6;
  bool check_command_is_important = 7;
  optional bool check_freshness = 8;  // Optional - Default value: 0
  optional uint32 check_interval = 9;  // Optional - Default value: 5
  string check_period = 10;
  repeated string contactgroups = 11;
  repeated string contacts = 12;
  repeated CustomVariable customvariables = 13;
  string display_name = 14;
  string event_handler = 15;
  optional bool event_handler_enabled = 16;  // Optional - Default value: true
  optional uint32 first_notification_delay = 17;  // Optional - Default value: 0
  optional bool flap_detection_enabled = 18;  // Optional - Default value: true
  optional uint32 flap_detection_options = 19;  // Optional
  optional uint32 freshness_threshold = 20;  // Optional - Default value: 0
  optional uint32 high_flap_threshold = 21;  // Optional - Default value: 0
  repeated string hostgroups = 22;
  repeated string hosts = 23;
  string icon_image = 24;
  string icon_image_alt = 25;
  optional uint32 initial_state = 26;  // Optional - Default value: engine::service::state_ok
  optional bool is_volatile = 27;  // Optional - Default value: false
  optional uint32 low_flap_threshold = 28;  // Optional - Default value: 0
  optional uint32 max_check_attempts = 29;  // Optional - Default value: 3
  string notes = 30;
  string notes_url = 31;
  optional bool notifications_enabled = 32;  // Optional - Default value: true
  optional uint32 notification_interval = 33;  // Optional - Default value: 30
  optional uint32 notification_options = 34;  // Optional
  optional string notification_period = 35;  // Optional
  optional bool obsess_over_service = 36;  // Optional - Default value: true
  optional bool process_perf_data = 37;  // Optional - Default value: true
  optional bool retain_nonstatus_information = 38;  // Optional - Default value: true
  optional bool retain_status_information = 39;  // Optional - Default value: true
  optional uint32 retry_interval = 40;  // Optional - Default value: 1
  optional uint32 recovery_notification_delay = 41;  // Optional
  repeated string servicegroups = 42;
  string service_description = 43;
  uint64 host_id = 44;
  uint64 service_id = 45;
  optional uint32 stalking_options = 46;  // Optional - Default value: service::none
  optional string timezone = 47;  // Optional
  optional uint64 severity_id = 48;  // Optional
  optional uint64 icon_id = 49;  // Optional
}

message Servicedependency {
  Object obj = 1;
  string dependency_period = 2;
  DependencyKind dependency_type = 3;
  repeated string dependent_hostgroups = 4;
  repeated string dependent_hosts = 5;
  repeated string dependent_servicegroups = 6;
  repeated string dependent_service_description = 7;
  optional uint32 execution_failure_options = 8;  // Optional
  repeated string hostgroups = 9;
  repeated string hosts = 10;
  optional bool inherits_parent = 11;  // Optional - Default value: false
  optional uint32 notification_failure_options = 12;  // Optional
  repeated string servicegroups = 13;
  repeated string service_description = 14;
}

message Serviceescalation {
  Object obj = 1;
  repeated string contactgroups = 2;
  optional uint32 escalation_options = 3;  // Optional - Default value: serviceescalation::none
  optional string escalation_period = 4;  // Optional
  optional uint32 first_notification = 5;  // Optional - Default value: -2
  repeated string hostgroups = 6;
  repeated string hosts = 7;
  optional uint32 last_notification = 8;  // Optional - Default value: -2
  optional uint32 notification_interval = 9;  // Optional - Default value: 0
  repeated string servicegroups = 10;
  repeated string service_description = 11;
}

message Serviceextinfo {
  Object obj = 1;
  string action_url = 2;
  string icon_image = 3;
  string icon_image_alt = 4;
  repeated string hostgroups = 5;
  repeated string hosts = 6;
  string notes = 7;
  string notes_url = 8;
  string service_description = 9;
}

message Servicegroup {
  Object obj = 1;
  string action_url = 2;
  string alias = 3;
  repeated PairString members = 4;
  string notes = 5;
  string notes_url = 6;
  uint32 servicegroup_id = 7;
  repeated string servicegroup_members = 8;
  string servicegroup_name = 9;
}

message Severity {
  Object obj = 1;
  KeyType key = 2;
  uint32 level = 3;
  uint64 icon_id = 4;
  string name = 5;
}

message Tag {
  Object obj = 1;
  KeyType key = 2;
  string name = 3;
}

message Timeperiod {
  Object obj = 1;
  string alias = 2;
  ExceptionArray exceptions = 3;
  repeated string exclude = 4;
  string timeperiod_name = 5;
  DaysArray timeranges = 6;
}
